ProjectCoimbra.UWP/Project.Coimbra/Pages/GamePage.xaml.cs

```diff
@@ -2,217 +2,343 @@
-                    this.midiEngine.Start();
-                    return;
-                }
-
-                var difference = UserData.StartTime.Value - currentTime;
-                this.Clockface.Text = difference.ToString("T", CultureInfo.CurrentCulture);
-            }
-            else
-            {
-                if (MultiPlayerData.StartTime == DateTime.MinValue)
-                {
-                    return;
-                }
-
-                if (MultiPlayerData.StartTime > DateTime.UtcNow)
-                {
-                    var difference = MultiPlayerData.StartTime - currentTime;
-                    this.Clockface.Text = difference.ToString(@"mm\:ss", CultureInfo.CurrentCulture);
-                }
-                else
-                {
-                    this.timer.Stop();
-                    this.Clockface.Text = string.Empty;
-                    this.InputControl.Pitches = UserData.PitchMap.Pitches;
-                    this.midiEngine.Start();
-                }
-            }
-        }
-    }
+		private void RemoveOldNotes()
+		{
+			foreach (var (key, value) in this.notesOnScreen.Where(pair => (DateTime.UtcNow - pair.Value.FirstDisplayed).TotalSeconds > 20))
+			{
+				this.InputControl.RemoveNote(value.NoteControl);
+				_ = this.notesOnScreen.Remove(key, out _);
+			}
+		}
+
+		private void RenderTimeToNextNote(TimeSpan currentTime)
+		{
+			MarkablePlaybackEvent lastNoteOnScreen = null;
+			if (this.notesOnScreen.Values.Count > 0)
+			{
+				lastNoteOnScreen = this.notesOnScreen.Values.Aggregate((a, b) => a.MarkablePlaybackEvent.RawTime > b.MarkablePlaybackEvent.RawTime ? a : b).MarkablePlaybackEvent;
+			}
+
+			TimeSpan lastNoteOffScreenTime = default(TimeSpan);
+
+			if (lastNoteOnScreen != null && lastNoteOnScreen.Time.TotalMilliseconds > 0)
+			{
+				lastNoteOffScreenTime = lastNoteOnScreen.Time + noteDuration;
+			}
+
+			long nextNoteTime = GetNextNoteTime((long)currentTime.TotalMilliseconds);
+			if (nextNoteTime < 0)
+			{
+				return;
+			}
+
+			long timeToNextNote = nextNoteTime - (long)currentTime.TotalMilliseconds;
+
+			long timeToNextNoteAfterLastNote = timeToNextNote -
+				// Substract note duration because it will show up earlier
+				(long)noteDuration.TotalMilliseconds +
+				// Substract currentTime to get the difference
+				(long)currentTime.TotalMilliseconds;
+
+			// No point of showing this if the next note is less than 3 second away
+			if ((lastNoteOffScreenTime.TotalMilliseconds == 0 || currentTime.TotalMilliseconds - lastNoteOffScreenTime.TotalMilliseconds > 1000) &&
+				(timeToNextNoteAfterLastNote > 3000 || previousTimeToNote >= 1000) &&
+				timeToNextNote >= 1000 &&
+				(lastPlayedNoteTimeIndex < this.noteTimes.Count - 1 ||
+				(lastPlayedNoteTimeIndex == this.noteTimes.Count - 1 && nextNoteTime > currentTime.TotalMilliseconds)))
+			{
+				previousTimeToNote = timeToNextNote;
+				this.dotCounter = 0;
+				var seconds = timeToNextNote / 1000;
```

Comment:
There are a lot of 1000s in this file. Consider extracting them to a named constant to make the intent clearer